torch/nn/parallel/_functions.py

```diff
@@ -51,12 +53,23 @@ def forward(ctx, target_device, dim, *inputs):
         ctx.target_device = target_device
         ctx.dim = dim
         ctx.input_gpus = tuple(map(lambda i: i.get_device(), inputs))
+        if all(t.dim() == 0 for t in inputs) and dim == 0:
```

Comment:
Yes it is intended.

Reply:
But isn't this contradictory to the configurable `dim` argument of `DataParallel` ? So if one has the samples at `dim=1` and passes 1 to `DataParallel`, everything goes well under this unsqueezing logic where it fails.

Reply:
:) Maybe I am missing something. I wanted to use DataParallel with my code on 0.4.1. My sample axis is the 1st axis so I use `dim=1` in order to scatter through that dimension. Everything works the workers receives the portions of the batches, computes the corresponding **scalar** losses. During gather however, i get an error about not being able to gather them. So I thought that this commit was intended to fix this by doing an unsqueeze but for my case since I give `dim=1` to DataParallel(), this unsqueezing does not happen for my case because of `dim == 0:` at the end of the if condition.

Shouldn't this part of the code be independent of the specific value of the `dim` ?

Reply:
Sounds like the problem is that `dim` is a single argument saying which dimension to split inputs and cat outputs on, but in your use case they should be different. Tbh I really dislike this special case for scalars... It makes the interface so much uglier...